### DISCUSSION TITLE
[@exp] rules unit testing

### DISCUSSION THREAD

**Discussion by laurentpayot** [firebase/firebase-js-sdk#4715] [Q&A] [answered]
Hi, so far using firebase@exp has been great (35% bundle size reduction :tada:) but it broke all my unit tests :confused: 

If I use the following `db`
```js
import * as firebase from '@firebase/rules-unit-testing'

const app = firebase.initializeTestApp({…})
export const db = app.firestore()
```
in (for instance)
```js
const batch = writeBatch(db)
```
I get
```
FirebaseError: Type does not match the expected instance. Did you pass a reference from a different Firestore SDK?
```

Should I open an issue in firebase-tools or is there a workaround?

**Comment by laurentpayot** [comment]
@akauppi except a couple of firebase events triggered cloud functions my app is totally client-side based.  So, client-side, I have CRUD functions for my entities that use firestore primitives. These CRUD functions have the database dependency injected as a parameter for easier testing.
When running my Jest tests I’m passing to my CRUD functions different `@firebase/rules-unit-testing` databases [initialized with their respective test user auth data](https://firebase.google.com/docs/firestore/security/test-rules-emulator#run_local_tests).
This way, with the current non-alpha release, I could run lots of tests to ensure that an entity created by user A cannot be modified by user B and basic security stuff like that.

But now I cannot inject firestore test databases anymore as their implementation is totally different from the firebase@exp firestore implementation I use on the client side.

Is it a clearer description?

**Reply by akauppi** [reply]
What I think is going on here is a collision between two separate but similar libraries. Let’s see it based on your first post:

```
import * as firebase from '@firebase/rules-unit-testing'

const app = firebase.initializeTestApp({…})
export const db = app.firestore()
```

All of this is based on `rules-unit-testing` which resembles the 8.x Firebase API but may have a completely different implementation. We won’t know, as users. Anyhow, it doesn’t even know the `@exp` API’s exist.

```
const batch = writeBatch(db)
```

Where did you get the `writeBatch` from? Presumably, it’s `@exp` and you are feeding it an object from the `rules-unit-testing` universe.

Do you see where you go off the road?  I don’t know why it used to work. Maybe the `rules-unit-testing` implementation is close enough with the 8.x API that this wasn’t breaking things. But - conceptually - I imagine also the earlier code was ”broken”.

**Comment by Feiyang1** [comment]
`@firebase/rules-unit-testing` depends on `firebase@8`, so the firestore instance you got is a v8 implementation, while `writeBatch()` expects an exp instance. 

We have some planned changes for `@firebase/rules-unit-testing` to address some of its current shortcomings and make it work with the new SDK. It's still being worked on, so unfortunately you can't use the exp package with the existing `@firebase/rules-unit-testing` until that comes out.

@yuchenshi Anything you want to add?

**Reply by akauppi** [reply]
@Feiyang1 Would you have any estimate on the `@firebase/rules-unit-testing` changes? A week? I was just today doing work on that, and considering bypassing it with the Firestore REST API. That would make rules testing detached from client libraries.

**Reply by Feiyang1** [reply]
We don't promise specific timelines, but I will say it's on our immediate roadmap.

**Reply by laurentpayot** [reply]
It’s been a looong month developing without touching existing Firestore stuff because of the lack of unit tests. Any news to share before [Google I/O 2021](https://events.google.com/io/session/5c6cbd69-90cf-4a7f-b8e2-9d79de61fb2c?lng=en)?

**Reply by akauppi** [reply]
> It’s been a looong month developing without touching existing Firestore stuff because of the lack of unit tests. Any news to share before [Google I/O 2021](https://events.google.com/io/session/5c6cbd69-90cf-4a7f-b8e2-9d79de61fb2c?lng=en)?

I am working on side support. Updating a library I use for testing so that it’s not depending on `rules-unit-testing` but has its own code. Will hopefully be available by Wed.

**Reply by yuchenshi** [reply]
Support for rules unit testing has been merged in #4837 and will be included in the next exp release. tl;dr you'll be able to specify a mock user directly through `useFirestoreEmulator(db, host, port, {mockUserToken: {sub: "myUid", ...}})`. You can still use some emulator utilities in `@firebase/rules-unit-testing` to update rules or clear data, but there is no need to `initializeTestApp` through another library so you won't get the version conflicts.

**Comment by yuchenshi** [comment, marked as answer]
Aug 2021 update: We've released [`@firebase/rules-unit-testing` v2](https://www.npmjs.com/package/@firebase/rules-unit-testing), which is a major revamp and works with v9 JS SDK (modular / compat).

See [our quickstarts](https://github.com/firebase/quickstart-testing/tree/master/unit-test-security-rules-v9) on using the new RUTv2 API (docs: [guide](https://firebase.devsite.corp.google.com/docs/rules/unit-tests#run_local_unit_tests_with_the_version_9_javascript_sdk) | [reference](https://firebase.devsite.corp.google.com/docs/reference/emulator-suite/rules-unit-testing/rules-unit-testing)). The instructions below are still possible for power users who wants to keep dependencies to a minimum (_but you're pretty much on your own if you choose this path_).

----

Earlier response:

As promised, `useFirestoreEmulator(db, host, port, {mockUserToken: {...}})` is now released. Try it out with the new Beta modular SDK and let us know what you think! `useDatabaseEmulator` also got a similar upgrade and storage is ~coming soon~ also released.

**Reply by akauppi** [reply]
What can I do with it? I.e. is there documentation?

**Reply by yuchenshi** [reply]
It's just got out and we're still working on documentation! In the meantime, here's a sneak peak on how you'd structure your code using the `mockUserToken` feature: https://github.com/firebase/quickstart-testing/pull/136/files#diff-8fd26621b78cde7a5153d88935007aa416084ad30903814e07adb3b28f734e15R81

(spoiler: it's almost the same, except you can use the modular v9 SDK directly instead of `initializeTestApp`.)

**Reply by laurentpayot** [reply]
@yuchenshi I know you’re still working on documentation but before I start the long task of rewriting my tests can you just tell us quickly how the @firebase/rules-unit-testing library is supposed to be used with that? (I’m using Jest so I don’t neeed `assertFails` nor `assertSucceeds`)

**Reply by akauppi** [reply]
@laurentpayot I tried to reach you over twitter. Since you are using Jest, can you use Jest 27, maybe? 

In such a case, I released this on Wed: http://npmjs.com/firebase-jest-testing

**Reply by laurentpayot** [reply]
@akauppi thanks I will git it a try. But for now I’m happy with my setup and I try to keep my dependencies to a bare minimum :wink:

**Reply by yuchenshi** [reply]
@laurentpayot As mentioned above, since you can mock tokens using the JS SDK itself, `@firebase/rules-unit-testing` will only be useful for utility functions and is not strictly required.

Just a friendly reminder that `assertFails` actually does more than you think -- it asserts that the promise must be rejected with "permission denied" instead of other code errors. You can of course replicate that yourself if you prefer to keep dependencies to a bear minimum -- https://github.com/firebase/quickstart-testing/pull/136/files#diff-8fd26621b78cde7a5153d88935007aa416084ad30903814e07adb3b28f734e15R106-R109 shows you exactly how. (The example uses `assert` but you can totally do the same thing with Jest APIs.)

Apart from that, you may find it handy to be able to clear your Firestore data and set security rules using the library, but again those are also [available via REST](https://firebase.google.com/docs/emulator-suite/connect_firestore#clear_your_database_between_tests). Just let us know what you think if you decide to go that route -- we're always open to improving the experience, be it code changes, documentation or examples.

**Reply by laurentpayot** [reply]
@yuchenshi I found out that `loadFirestoreRules` from `@firebase/rules-unit-testing` is the only way to update Firestore rules after they have been modified.
It seems that [`releaseFirestoreRulesetFromSource`](https://github.com/firebase/quickstart-testing/pull/136/files#diff-8fd26621b78cde7a5153d88935007aa416084ad30903814e07adb3b28f734e15R57) doesn’t work and raises security concerns so for now I’m super happy with `loadFirestoreRules` :wink:

**Reply by yuchenshi** [reply]
@laurentpayot: We hear you regarding the confusion for loading Firestore rules, and we've now made it a built-in option for  [`@firebase/rules-unit-testing` v2](https://www.npmjs.com/package/@firebase/rules-unit-testing). See code example here: https://github.com/firebase/quickstart-testing/pull/209/files#diff-6f539bff6bb3d5eeb2d2074b7d97c417466274d42384b3b8c504f94633ea3393R33

We also packed RUTv2 with utils to help reduce boilerplate while you rewrite your tests. More docs coming soon to make the migration smoother.

**Reply by laurentpayot** [reply]
@yuchenshi RUT v2 is great, authentication contexts make a lot of sense.
In the code example you gave, I noticed two things:
- When running multiple test files in parallel (with [Jest](https://jestjs.io/)), `projectId` option of `initializeTestEnvironment()` must be filled. Otherwise all test files will use the same default database/app and a test and will clear this shared database while other tests are running.
- Whith TypeScript, when injecting a RUT Firestore database as a dependency of my CRUD Firestore functions I was getting:  `Type 'Firestore' is missing the following properties from type 'Firestore': type, toJSONts(2322)`. So I had to do a type assertion `<Firestore><unknown>`.

**Reply by yuchenshi** [reply]
@laurentpayot Your first point is great and I've tracked it internally for documentation. The second issue should be solved by upgrading both `@firebase/rules-unit-testing` and the `firebase` package to the latest version. If you somehow have two `package.json`s (e.g. one for your web app and one for tests), make sure both are updated and in sync. Please file a bug if you still see this after upgrading.

**Reply by laurentpayot** [reply]
@yuchenshi I beleive I am using the latest `@firebase/rules-unit-testing` and `firebase` versions. Here is an extract from my `package.json` file:
```json
{
  "devDependencies": {
    "@firebase/rules-unit-testing": "^2.0.0",
    "firebase": "^9.0.1",
    "firebase-admin": "^9.100.0-alpha.1",
    "firebase-functions": "^3.15.5",
    "firebase-functions-test": "^0.3.2",
    "firebase-tools": "^9.17.0",
  }
}
```
When I click on "Go to Type Definition" of VS Code from the result of `testEnv.authenticatedContext(...).firestore()` I land into `node_modules/firebase/compat/index.d.ts`, inside a `firebase.firestore` module. And there the exported `Firestore` class has no `type` attribute nor `toJSON` method.
What is the compat library used for?

**Reply by ceolinwill** [reply]
@laurentpayot Did you manage to find a solution for these type issues? I just upgraded a project to the latest versions of `firebase` and `@firebase/rules-unit-testing` but I'm getting type errors too when I try to follow the [docs](https://firebase.google.com/docs/rules/unit-tests#run_local_unit_tests_with_the_version_9_javascript_sdk).

**Reply by laurentpayot** [reply]
@wceolin As mentioned above,  the only solution I found to calm TypeScript down was to do a type assertion like so:
```typescript
import type { Firestore } from '@firebase/firestore'

const aliceDb = <Firestore><unknown>testEnv.authenticatedContext('alice').firestore()
```

**Reply by yuchenshi** [reply]
@laurentpayot @wceolin We've located the issue and is working on a fix now (https://github.com/firebase/firebase-js-sdk/pull/5500). Thanks a lot for your help!

Technical details on this: Firebase JS SDK v9 has a compatibility layer that allows v8-like non-modular APIs. The rules unit testing library is supposed to work with either that or the modular v9 APIs but due to a typing bug earlier TypeScript complains for the latter. #5500 fixes that, and type assertions are valid workarounds until the fix is released soon.

**Reply by laurentpayot** [reply]
The [#5550 "solution"](https://github.com/firebase/firebase-js-sdk/issues/5550#issuecomment-928453315) does calm TypeScript in the editor, but when actually running my tests with the test `db` injected in my front end code I get the same type errors as mentioned above :neutral_face:
So I have to keep type assertions for now…

**Reply by yuchenshi** [reply]
@laurentpayot See https://github.com/firebase/firebase-js-sdk/issues/5550#issuecomment-928453315. BTW, I'd really appreciate it if we can continue any further discussion on relevant issues instead of this thread.

**Reply by laurentpayot** [reply]
@yuchenshi the link I was reffering to in my previous comment is the one you want me to see.
Sure, I think the object of this discussion is now discussed enough. Thanks for all your time. :+1:

**Comment by akauppi** [comment]
@yuchenshi Nice that Rules v2 is getting attention.

You may want to consider two minor things about the tests you linked to, before releasing:

1. https://github.com/firebase/quickstart-testing/pull/209/files#diff-6f539bff6bb3d5eeb2d2074b7d97c417466274d42384b3b8c504f94633ea3393R64-R68

   ..contains setup within tests themselves, which is not giving a good example. Rather, move those to `beforeEach`?

2. Instead of plain `db` for the database handle that acts with Alice's rights, you could name it `dbAlice`.

Why do these matter? For one, any sample you ship is educational to the larger audience, some of whom are novices and will simply copy-paste what's out there. 

Second, tests should be even clearer and more readable than code, since they can be used also by managers and such to discuss about features - and see what's being tested.

Great that Firebase has good testing tools.

**Reply by rachelmyers** [reply]
@akauppi, thanks for spending so much time on this; it's clear you're invested and want great code examples. The PR will be updated before it's merged, but we can't promise we will or won't make those changes. 

1. Setup within the test vs in a `beforeEach` depends on the test file. If each test setup is pretty unique, I have no problem putting the setup in each test. We'll have to see what the final state of the file is. 
2. We use both `db` and making the name specific to a user in our samples. It depends on the other tests in the file. Again, it depends on the final state of this test suite.
